### DISCUSSION TITLE
Orchard Core Headless: APIs to create, update or delete a Content Item instance

### DISCUSSION THREAD

**Discussion by MarGraz** [OrchardCMS/OrchardCore#15372] [Q&A] [answered]
Hi,

I'm trying to learn more about the APIs available in Orchard Core, especially regarding GraphQL. I'm interested in whether it's possible to **create, update, or delete a Content Item instance through APIs**. 

I've looked at the [documentation](https://docs.orchardcore.net/fr/latest/docs/reference/modules/Apis.GraphQL/), but the details on how it works aren't clear to me. For updating and deleting a Content Items instance, do I need to create [my own ApiController](https://docs.orchardcore.net/projects/O1/en/latest/Documentation/WebApi-In-Orchard/), **or does Orchard Core, with its dynamic types, already provide a way to handle these operations?** 

Could you suggest how I might create, update, or delete a `Page` Content Item **instance** through the API endpoint? Additionally, is there a method to view all available endpoints?

Thank you

**Comment by MarGraz** [comment, marked as answer]
Looking at the Orchard Core solution, and going deeper into the _issues_ (not the _discussions_), I found a useful issue [here](https://github.com/OrchardCMS/OrchardCore/issues/8656) that is similar to my case.

In the end, **the module** `OrchardCore.Contents` **exposes an** `ApiController.cs` with [endpoints](https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.Contents/Controllers/ApiController.cs) to **create, update, and delete** _content items_ created in the "Admin" backoffice using _Postman_ (and the OpenID Connect module for the authentication).

@Skrypt, I mention you because you also replied to the [issue](https://github.com/OrchardCMS/OrchardCore/issues/8656) I mentioned before. Can I add information about these endpoints in the documentation? Because I think it's a significant omission. 

Is it appropriate to add this information to the [Content Types](https://docs.orchardcore.net/fr/latest/docs/reference/modules/ContentTypes/) docs?

Thank you

**Reply by MarGraz** [reply]
@Piedone I also want to ask you if I can add information about he module `OrchardCore.Contents` [ApiController.cs](https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.Contents/Controllers/ApiController.cs) endpoints to create, update, and delete content items. Because I think it's a significant omission.

Could be also nice to add a link to this fantastic [Swagger module](https://github.com/OrchardCoreContrib/OrchardCoreContrib.Modules/blob/main/src/OrchardCoreContrib.Apis.Swagger/README.md). 

Is it appropriate to add this information to the [Content Types](https://docs.orchardcore.net/fr/latest/docs/reference/modules/ContentTypes/) docs? 🤔

**Reply by Piedone** [reply]
That would be great! I think this is the most fitting page, since that controller is under the Contents module: https://docs.orchardcore.net/en/latest/docs/reference/modules/Contents/.

Please prefer the English docs, since that's the one accessible to the most people.

**Reply by hishamco** [reply]
@MarGraz let me know your experience in using Swagger Module, what's good & bad

**Reply by Piedone** [reply]
Also, feel free to request review from me when you have the docs page ready. And see: https://docs.orchardcore.net/en/latest/docs/topics/docs-contributions/.

**Reply by MarGraz** [reply]
@hishamco I tried the "Swagger Module", it's really simple to install and use. I started my project using a "blank site" recipe, so I can only see the "Content" endpoints. 

Is there a "User" endpoint? 
Because I'm looking for a way to list all the users, but it seems that the "User Module" doesn't have any [ApiController](https://github.com/OrchardCMS/OrchardCore/tree/main/src/OrchardCore.Modules/OrchardCore.Users/Controllers) to do this. I opened a discussion [here](https://github.com/OrchardCMS/OrchardCore/discussions/15373).

Thank you

**Reply by hishamco** [reply]
I prefer to file an issue or open a discussion to OCC repo

**Reply by MarGraz** [reply]
Yes, I already I opened a discussion [here](https://github.com/OrchardCMS/OrchardCore/discussions/15373).

**Reply by hishamco** [reply]
I was referring to Swagger Module :)

**Reply by MarGraz** [reply]
@hishamco I have opened a [new discussion]( https://github.com/OrchardCoreContrib/OrchardCoreContrib.Modules/issues/116) in the OCC repo 😊

**Comment by MarGraz** [comment]
Using _Postman_, **it's possible to call the** `ApiController.cs` endpoints after configuring the "OpenID Connect" module. I tested this using the "[Try Orchard](https://try.orchardcore.net/)" website. **The authentication** process **is successful**, and I receive a valid OAuth 2 token. **But, when attempting to use the "Admin" user**, which is supposed to have the highest privileges, **I cannot perform GET, UPDATE, or DELETE operations**, I receive a 401 unauthorized error. **Instead, if I grant all privileges to the "anonymous" user role** to call the API, and remove the OAuth token, it works 🤔

Could this issue be related to the user not having the [permission at line 43](https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.Contents/Controllers/ApiController.cs)? However, my user has the `Permissions.AccessContentApi`.

This is the part at line 43:

 ```csharp
 [Route("{contentItemId}"), HttpGet]
        public async Task<IActionResult> Get(string contentItemId)
        {
            if (!await _authorizationService.AuthorizeAsync(User, Permissions.AccessContentApi))
            {
                return this.ChallengeOrForbid("Api");
            }
            
            [...]
            
          }
```

**Any suggestions?**

**Reply by hishamco** [reply]
I prefer to file a new issue describing the steps to reproduce, attaching Postman collection will be a plus

**Reply by MarGraz** [reply]
@hishamco I have created a new issue [here](https://github.com/OrchardCMS/OrchardCore/issues/15381). I have also attached the Postman collection. I hope that the "try.orchardcore" site will be not reset soon.

**Reply by Piedone** [reply]
Also see: https://github.com/Lombiq/Orchard-Core-API-Client.

**Reply by hishamco** [reply]
Exactly, I forgot to mention it, @Piedone you remind me with another Zolte :)

**Reply by Piedone** [reply]
There are too many Zoltáns in Lombiq :D.

**Reply by hishamco** [reply]
BTW Zoltan is used in Arabic as "Sultan" :)

**Reply by Piedone** [reply]
Yep, that's where it comes from :).

**Comment by hishamco** [comment]
> Could you suggest how I might create, update, or delete a Page Content Item instance through the API endpoint? Additionally, is there a method to view all available endpoints?

Could you try Swagger Module https://github.com/OrchardCoreContrib/OrchardCoreContrib.Modules/blob/main/src/OrchardCoreContrib.Apis.Swagger/README.md

**Reply by MarGraz** [reply]
@hishamco, oh, that's great 🥳 Who is maintaining the `OrchardCoreContrib.Modules`? I think it could also be useful to mention it in the documentation.

Please, also see my previous two replies. I think we posted at the same time, and I'm asking about which part of the documentation to update 🙂

**Reply by hishamco** [reply]
> @hishamco, oh, that's great 🥳 Who is maintaining the OrchardCoreContrib.Modules? I think it could also be useful to mention it in the documentation.

It's me for now :) but all the Orchard Core contributors are welcome
